File: .a-new-structure/b-word/L1-omnicode/hybrid/pragma/dispatch.go
// #!omni code --go -library
// #!omni meta.key = B-L1-omnicode-hybrid-pragma-dispatch
// #!omni meta.from = B-L1-omnicode-hybrid-pragma
// #!omni meta.at = a-01.00
//omni:code --go -library
//omni:key B-L1-omnicode-hybrid-pragma-dispatch
//omni:version a-01.00
// ============================================================================
// METADATA
// ============================================================================
//
// OmniCode Format Dispatcher — Route Files to the Right Parser
//
// Key:     B-L1-omnicode-hybrid-pragma-dispatch
// Purpose: Read pragma, dispatch to L0 config/util parser (TOML, JSONC, JSON)
// Biblical: Genesis 1:4 - "And God divided the light from the darkness"
//           Separation enables identity — each format gets its proper handler.
// Authors: Nova Dawn
// Version: a-01.00
// Created: 2026-02-13
//
// The dispatcher bridges L1 (OmniCode identity) and L0 (format parsing).
// An .omni file with --toml format gets its pragma read by L1, then its
// content parsed by L0's TOML loader. An .omni file with --json gets L0's
// JSONC parser. An .omni file with --omni gets the native parser (future).
//
// ============================================================================
// END METADATA
// ============================================================================

package pragma

// ============================================================================
// SETUP
// ============================================================================

import (
	"fmt"
	"path/filepath"
	"sync"

	// L0 shared config utilities — format parsers, structured errors
	"creativeworkzstudio.com/bereshit/L0-universal/hybrid/config/util"
)

// --- File Extension Registry ---

// KnownExtensions maps file extensions to their OmniCode significance.
var KnownExtensions = map[string]string{
	".omni": "OmniCode native file",
	".ofd":  "OmniFold folder metadata",
	".toml": "TOML config (may contain pragma)",
	".json": "JSON data (may contain pragma in adjacent .omni)",
	".jsonc": "JSONC config (may contain pragma in adjacent .omni)",
}

// --- Dispatch Cache ---
//
// Caches LoadConfigMap results by filepath. Generic LoadConfig[T] can't be
// easily cached (type erasure), but the map variant is the common hot path
// for OmniFold traversal and tools that inspect arbitrary files.

type configMapEntry struct {
	pragma *Pragma
	data   map[string]any
}

var (
	dispatchMu    sync.RWMutex
	dispatchCache = make(map[string]*configMapEntry)
)

// ============================================================================
// END SETUP
// ============================================================================

// ============================================================================
// BODY
// ============================================================================

// ────────────────────────────────────────────────────────────────
// Helpers
// ────────────────────────────────────────────────────────────────

// IsOmniCodeFile returns true if the file extension indicates an OmniCode file.
func IsOmniCodeFile(path string) bool {
	ext := filepath.Ext(path)
	return ext == ExtOmni || ext == ExtOmniFold
}

// ────────────────────────────────────────────────────────────────
// Format Dispatch
// ────────────────────────────────────────────────────────────────

// LoadConfig reads an OmniCode-wrapped config file.
// Parses the pragma to determine format, then dispatches to the right L0 parser.
//
// Supports:
//   - .omni files with --toml format → L0 TOML loader
//   - .omni files with --json format → L0 JSONC loader
//   - .ofd files (folder metadata) → L0 JSONC or TOML loader based on pragma
//
// For pure OmniCode (--omni), returns an error — native parser not yet implemented.
//
// Uses L0 structured error types (util.LoadError) for consistent error handling
// across the L0-L1 boundary.
func LoadConfig[T any](path string) (*Pragma, *T, error) {
	// Step 1: Read pragma (L1 — identity, cached by Parse())
	p, err := Parse(path)
	if err != nil {
		return nil, nil, util.NewLoadError(path, "pragma", err)
	}

	// Step 2: Dispatch to L0 parser based on format
	var result *T

	switch p.Format {
	case FormatTOML:
		result, err = loadTOMLContent[T](path)
	case FormatJSON, FormatJSONC:
		result, err = util.LoadJSONCFile[T](path)
	case FormatOmni:
		return p, nil, util.NewLoadError(path, "dispatch",
			fmt.Errorf("native OmniCode parser not yet implemented"))
	default:
		return p, nil, util.NewLoadError(path, "dispatch",
			fmt.Errorf("unsupported config format %q", p.Format))
	}

	if err != nil {
		return p, nil, util.NewLoadError(path, "parse", err)
	}

	return p, result, nil
}

// LoadConfigMap loads an OmniCode config file into a generic map.
// Useful when the exact structure isn't known at compile time.
//
// Results are cached by filepath — subsequent calls for the same path
// return the cached result without disk I/O. Use InvalidateDispatchCache()
// after file modifications.
func LoadConfigMap(path string) (*Pragma, map[string]any, error) {
	// Normalize path for consistent cache keys
	path = filepath.Clean(path)

	// Check cache first (read lock — concurrent readers allowed)
	dispatchMu.RLock()
	if cached, ok := dispatchCache[path]; ok {
		dispatchMu.RUnlock()
		return cached.pragma, cached.data, nil
	}
	dispatchMu.RUnlock()

	// Cache miss — load via generic path
	p, result, err := LoadConfig[map[string]any](path)
	if err != nil {
		return p, nil, err
	}

	var data map[string]any
	if result != nil {
		data = *result
	}

	// Store in cache (write lock — exclusive access)
	dispatchMu.Lock()
	dispatchCache[path] = &configMapEntry{pragma: p, data: data}
	dispatchMu.Unlock()

	return p, data, nil
}

// ────────────────────────────────────────────────────────────────
// Format-Specific Loaders
// ────────────────────────────────────────────────────────────────

// loadTOMLContent loads TOML content from an OmniCode-wrapped file.
// Skips pragma lines (they're OmniCode, not TOML) and parses the rest.
//
// NOTE: For files where pragma is in TOML comments (# #!omni ...),
// the TOML parser naturally ignores comments. For bare pragma files (.omni),
// we need to strip pragma lines before parsing — this is the OmniCode-native
// case where pragma IS syntax, not comments.
//
// TODO(L1): Handle bare .omni files where pragma lines aren't comments.
// Current implementation works for .toml files with pragma in # comments.
func loadTOMLContent[T any](path string) (*T, error) {
	// For now, the TOML parser handles # #!omni lines as comments.
	// This works for .toml files. For .omni files with --toml format,
	// we'll need to strip pragma lines first (future L1 work).
	return util.LoadTOMLFile[T](path)
}

// ────────────────────────────────────────────────────────────────
// OmniFold (.ofd) Support
// ────────────────────────────────────────────────────────────────

// LoadFolder reads an OmniFold (.ofd) folder metadata file.
// Returns the pragma and raw content as a map.
func LoadFolder(path string) (*Pragma, map[string]any, error) {
	if ext := filepath.Ext(path); ext != ExtOmniFold && ext != ExtOmni {
		return nil, nil, util.NewLoadError(path, "dispatch",
			fmt.Errorf("expected %s or %s extension, got %s", ExtOmniFold, ExtOmni, ext))
	}

	p, data, err := LoadConfigMap(path)
	if err != nil {
		return nil, nil, err // Already a LoadError from LoadConfig
	}

	if p.FileType != TypeFolder {
		return p, data, util.NewLoadError(path, "validate",
			fmt.Errorf("expected folder type, got %s", p.FileType))
	}

	return p, data, nil
}

// ────────────────────────────────────────────────────────────────
// Cache Invalidation
// ────────────────────────────────────────────────────────────────

// InvalidateDispatchCache clears the entire dispatch cache.
// Call after bulk file modifications or during testing.
// Also clears the pragma cache since dispatch depends on it.
func InvalidateDispatchCache() {
	dispatchMu.Lock()
	dispatchCache = make(map[string]*configMapEntry)
	dispatchMu.Unlock()

	// Pragma cache is upstream — clear it too
	InvalidatePragmaCache()
}

// InvalidateDispatchCacheEntry removes a single entry from the dispatch cache.
// Also clears the corresponding pragma cache entry.
func InvalidateDispatchCacheEntry(path string) {
	path = filepath.Clean(path)
	dispatchMu.Lock()
	delete(dispatchCache, path)
	dispatchMu.Unlock()

	InvalidatePragmaCacheEntry(path)
}

// ============================================================================
// END BODY
// ============================================================================

// ============================================================================
// CLOSING
// ============================================================================
//
// The dispatch chain:
//
//   File (.omni/.ofd)
//     ↓ Parse pragma (L1)
//   Pragma { type, format, key, from, at }
//     ↓ Dispatch by format
//   L0 parser (TOML / JSONC / future: native OmniCode)
//     ↓ Returns typed data
//   Caller gets Pragma + parsed config
//
// This wires OmniCode identity into the config loading chain.
// The pragma tells us WHAT the file is. The L0 parsers know HOW to read
// the underlying format. L1 connects the two.
//
// Future: When the native OmniCode parser exists (L1 frontend/lexer),
// FormatOmni dispatch will route there instead of returning an error.
//
// ============================================================================
// END CLOSING
// ============================================================================

File: .a-new-structure/b-word/L1-omnicode/hybrid/pragma/src/dispatch.rs
//! OmniCode Format Dispatcher — Route Files to the Right Parser.
//!
//! **THIS IS THE L0-L1 BRIDGE.** Reads pragma (L1 domain), dispatches to
//! L0 format parsers (TOML, JSONC). The compiler witnessing this code
//! compile proves the layer contract is real.
//!
//! Genesis 1:4 — "And God divided the light from the darkness."
//! Separation enables identity — each format gets its proper handler.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};
use std::sync::{LazyLock, RwLock};

use serde::de::DeserializeOwned;

use bereshit_l0_config as config;

use crate::reader;
use crate::types::*;

// ────────────────────────────────────────────────────────────────
// Dispatch Cache
// ────────────────────────────────────────────────────────────────

/// Cached dispatch result for the map variant.
///
/// Generic `load_config<T>` can't be cached (type erasure), but
/// `load_config_map` is the common hot path for OmniFold traversal
/// and tools that inspect arbitrary files.
#[derive(Clone)]
struct ConfigMapEntry {
    pragma: Pragma,
    data: serde_json::Map<String, serde_json::Value>,
}

/// Global dispatch cache. RwLock for concurrent reads, exclusive writes.
static DISPATCH_CACHE: LazyLock<RwLock<BTreeMap<PathBuf, ConfigMapEntry>>> =
    LazyLock::new(|| RwLock::new(BTreeMap::new()));

// ────────────────────────────────────────────────────────────────
// Format Dispatch
// ────────────────────────────────────────────────────────────────

/// Load an OmniCode-wrapped config file.
///
/// Parses the pragma (L1) to determine format, then dispatches to the
/// right L0 parser. Returns both the pragma identity and the parsed data.
///
/// **L0 Connections (compiler-verified):**
///   - `Format::Toml` → `bereshit_l0_config::load_toml_file::<T>(path)`
///   - `Format::Json` → `bereshit_l0_config::load_jsonc_file::<T>(path)`
///   - `Format::Omni` → error (native parser not yet implemented)
///
/// Go equivalent: `func LoadConfig[T any](path string) (*Pragma, *T, error)`
pub fn load_config<T: DeserializeOwned>(path: &Path) -> Result<(Pragma, T), DispatchError> {
    // Step 1: Read pragma (L1 — identity)
    let pragma = reader::parse(path)?;

    // Step 2: Dispatch to L0 parser based on format
    // JSON routes to JSONC loader — JSONC is a superset of JSON
    let result: T = match pragma.format {
        Format::Toml => config::load_toml_file(path)?,
        Format::Json | Format::Jsonc => config::load_jsonc_file(path)?,
        Format::Omni => return Err(DispatchError::NativeNotImplemented),
        other => return Err(DispatchError::UnsupportedFormat(other)),
    };

    Ok((pragma, result))
}

/// Load an OmniCode config file into a generic map.
///
/// Useful when the exact structure isn't known at compile time.
///
/// Results are cached by filepath — subsequent calls for the same path
/// return the cached result without disk I/O. Use [`invalidate_dispatch_cache()`]
/// after file modifications.
///
/// Go equivalent: `func LoadConfigMap(path string) (*Pragma, map[string]interface{}, error)`
pub fn load_config_map(
    path: &Path,
) -> Result<(Pragma, serde_json::Map<String, serde_json::Value>), DispatchError> {
    // Normalize path — prevents cache misses from different representations.
    let canonical = std::fs::canonicalize(path).unwrap_or_else(|_| path.to_path_buf());

    // Check cache first (read lock)
    {
        let guard = DISPATCH_CACHE.read().unwrap_or_else(|e| e.into_inner());
        if let Some(entry) = guard.get(&canonical) {
            return Ok((entry.pragma.clone(), entry.data.clone()));
        }
    }

    // Cache miss — load via generic path
    let (pragma, data): (Pragma, serde_json::Map<String, serde_json::Value>) = load_config(path)?;

    // Store in cache (write lock) — keyed by canonical path
    {
        let mut guard = DISPATCH_CACHE.write().unwrap_or_else(|e| e.into_inner());
        guard.insert(
            canonical,
            ConfigMapEntry {
                pragma: pragma.clone(),
                data: data.clone(),
            },
        );
    }

    Ok((pragma, data))
}

// ────────────────────────────────────────────────────────────────
// OmniFold (.ofd) Support
// ────────────────────────────────────────────────────────────────

/// Load an OmniFold (.ofd) folder metadata file.
///
/// Returns the pragma and raw content as a map. Validates that
/// the file type is `folder`.
///
/// Go equivalent: `func LoadFolder(path string) (*Pragma, map[string]interface{}, error)`
pub fn load_folder(
    path: &Path,
) -> Result<(Pragma, serde_json::Map<String, serde_json::Value>), DispatchError> {
    let ext = path
        .extension()
        .and_then(|e| e.to_str())
        .map(|e| format!(".{e}"))
        .unwrap_or_default();

    if ext != EXT_OMNIFOLD && ext != EXT_OMNI {
        return Err(DispatchError::UnsupportedFormat(Format::Omni));
    }

    let (pragma, data) = load_config_map(path)?;

    if pragma.file_type != FileType::Folder {
        // Return data but log warning — Go version returns error
        return Err(DispatchError::UnsupportedFormat(pragma.format));
    }

    Ok((pragma, data))
}

// ────────────────────────────────────────────────────────────────
// Helpers
// ────────────────────────────────────────────────────────────────

/// Whether a file path has an OmniCode extension.
///
/// Go equivalent: `func IsOmniCodeFile(path string) bool`
pub fn is_omnicode_file(path: &Path) -> bool {
    crate::extensions::is_omnicode_file(path)
}

// ────────────────────────────────────────────────────────────────
// Cache Invalidation
// ────────────────────────────────────────────────────────────────

/// Clear the entire dispatch cache.
/// Also clears the pragma cache since dispatch depends on it.
///
/// Go equivalent: `func InvalidateDispatchCache()`
pub fn invalidate_dispatch_cache() {
    {
        let mut guard = DISPATCH_CACHE.write().unwrap_or_else(|e| e.into_inner());
        guard.clear();
    }
    // Pragma cache is upstream — clear it too
    crate::cache::invalidate_all();
}

/// Remove a single entry from the dispatch cache.
/// Also clears the corresponding pragma cache entry.
///
/// Go equivalent: `func InvalidateDispatchCacheEntry(path string)`
pub fn invalidate_dispatch_entry(path: &Path) {
    let canonical = std::fs::canonicalize(path).unwrap_or_else(|_| path.to_path_buf());
    {
        let mut guard = DISPATCH_CACHE.write().unwrap_or_else(|e| e.into_inner());
        guard.remove(&canonical);
    }
    crate::cache::invalidate(&canonical);
}

// ────────────────────────────────────────────────────────────────
// Tests
// ────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    fn test_dir(name: &str) -> std::path::PathBuf {
        let ts = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_nanos();
        let dir = std::env::temp_dir().join(format!("bereshit-l1-dispatch-{name}-{ts}"));
        fs::create_dir_all(&dir).unwrap();
        dir
    }

    fn write_file(root: &Path, rel: &str, content: &str) {
        let path = root.join(rel);
        fs::create_dir_all(path.parent().unwrap()).unwrap();
        fs::write(path, content).unwrap();
    }

    #[test]
    fn test_load_config_toml() {
        let dir = test_dir("toml");
        write_file(
            &dir,
            "config.toml",
            "# #!omni data --toml\n# #!omni meta.key = test-config\n\n[section]\nkey = \"value\"\n",
        );

        #[derive(serde::Deserialize)]
        struct Outer {
            section: Section,
        }
        #[derive(serde::Deserialize)]
        struct Section {
            key: String,
        }

        let (pragma, data): (Pragma, Outer) = load_config(&dir.join("config.toml")).unwrap();
        assert_eq!(pragma.format, Format::Toml);
        assert_eq!(pragma.key, "test-config");
        assert_eq!(data.section.key, "value");

        let _ = fs::remove_dir_all(&dir);
    }

    #[test]
    fn test_load_config_jsonc() {
        let dir = test_dir("jsonc");
        write_file(
            &dir,
            "config.jsonc",
            "// #!omni data --json\n// #!omni meta.key = test-json\n{\n  // comment\n  \"name\": \"bereshit\",\n}\n",
        );

        #[derive(serde::Deserialize)]
        struct Config {
            name: String,
        }

        let (pragma, data): (Pragma, Config) = load_config(&dir.join("config.jsonc")).unwrap();
        assert_eq!(pragma.format, Format::Json);
        assert_eq!(pragma.key, "test-json");
        assert_eq!(data.name, "bereshit");

        let _ = fs::remove_dir_all(&dir);
    }

    #[test]
    fn test_load_config_omni_not_implemented() {
        let dir = test_dir("omni-native");
        write_file(&dir, "file.omni", "#!omni code\n#!omni meta.key = native\n");

        let err = load_config::<serde_json::Value>(&dir.join("file.omni")).unwrap_err();
        assert!(matches!(err, DispatchError::NativeNotImplemented));

        let _ = fs::remove_dir_all(&dir);
    }

    #[test]
    fn test_load_config_unsupported_format() {
        let dir = test_dir("unsupported");
        write_file(&dir, "file.omni", "#!omni code --go\n");

        let err = load_config::<serde_json::Value>(&dir.join("file.omni")).unwrap_err();
        assert!(matches!(err, DispatchError::UnsupportedFormat(Format::Go)));

        let _ = fs::remove_dir_all(&dir);
    }

    #[test]
    fn test_load_config_missing_pragma() {
        let dir = test_dir("no-pragma");
        write_file(&dir, "plain.toml", "[section]\nkey = \"value\"\n");

        let err = load_config::<serde_json::Value>(&dir.join("plain.toml")).unwrap_err();
        assert!(matches!(err, DispatchError::Pragma(_)));

        let _ = fs::remove_dir_all(&dir);
    }

    #[test]
    fn test_load_config_map_toml() {
        let dir = test_dir("map-toml");
        write_file(
            &dir,
            "data.toml",
            "# #!omni data --toml\n# #!omni meta.key = map-test\n\n[info]\nname = \"test\"\n",
        );

        let (pragma, data) = load_config_map(&dir.join("data.toml")).unwrap();
        assert_eq!(pragma.key, "map-test");
        let info = data.get("info").expect("should have info key");
        assert!(info.is_object());

        let _ = fs::remove_dir_all(&dir);
    }
}
